master/internal/db/postgres_trial.go

```diff
@@ -468,99 +468,133 @@ func (db *PgDB) updateLatestValidationID(ctx context.Context, tx *sqlx.Tx, trial
+	if err != nil {
+		return 0, errors.Wrap(err, "checking for metric rollbacks")
+	}
+	return int(affectedRows), nil
+}
+
+func (db *PgDB) addRawMetrics(ctx context.Context, tx *sqlx.Tx, metricsBody *map[string]interface{},
+	runID, trialID, lastProcessedBatch int32,
+	pType model.MetricPartitionType, mType *string,
+) (int, error) {
+	var metricRowID int
+	if err := tx.QueryRowContext(ctx, `
+INSERT INTO metrics
+	(trial_id, trial_run_id, end_time, metrics, total_batches, partition_type, custom_type)
+VALUES
+	($1, $2, now(), $3, $4, $5, $6)
+RETURNING id`,
+		trialID, runID, *metricsBody, lastProcessedBatch, pType, mType, // CHECK: are nulls handled?
+	).Scan(&metricRowID); err != nil {
+		return metricRowID, errors.Wrap(err, "inserting metrics")
+	}
+
+	return metricRowID, nil
+}
+
 // AddTrialMetrics inserts a set of trial metrics to the database.
 func (db *PgDB) addTrialMetrics(
-	ctx context.Context, m *trialv1.TrialMetrics, isValidation bool,
-) (rollbacks map[string]int, err error) {
-	rollbacks = make(map[string]int)
-	trialMetricTables := []string{"raw_steps", "raw_validations"}
-	targetTable := "raw_steps"
+	ctx context.Context, m *trialv1.TrialMetrics, pType model.MetricPartitionType,
+	mType *string,
+) (rollbacks int, err error) {
+	/*
+		TODO(hamid):
+		- MetricType could live in trialv1.TrialMetrics :thinking_face:
+	*/
+	isValidation := pType == model.ValidationMetric
+
 	metricsJSONPath := "avg_metrics"
 	metricsBody := map[string]interface{}{
-		"avg_metrics":   m.Metrics.AvgMetrics,
+		metricsJSONPath: m.Metrics.AvgMetrics,
 		"batch_metrics": m.Metrics.BatchMetrics,
 	}
 	if isValidation {
 		metricsJSONPath = "validation_metrics"
-		targetTable = "raw_validations"
 		metricsBody = map[string]interface{}{
-			"validation_metrics": m.Metrics.AvgMetrics,
+			metricsJSONPath: m.Metrics.AvgMetrics,
 		}
 	}
 	return rollbacks, db.withTransaction("add training metrics", func(tx *sqlx.Tx) error {
 		if err := checkTrialRunID(ctx, tx, m.TrialId, m.TrialRunId); err != nil {
 			return err
 		}
 
-		rollbackHappened := false
-		for _, table := range trialMetricTables {
-			comparator := ">"
-			if table == targetTable {
-				// we mark metrics reported in the same table with the same batch number
-				// as the metric being added as `archived`.
-				comparator = ">="
-			}
-			res, err := tx.ExecContext(ctx, fmt.Sprintf(`
-UPDATE %s SET archived = true
-WHERE trial_id = $1
-  AND archived = false
-  AND trial_run_id < $2
-  AND total_batches %s $3;
-	`, table, comparator), m.TrialId, m.TrialRunId, m.StepsCompleted)
-			if err != nil {
-				return errors.Wrap(err, "archiving metrics")
-			}
-			affectedRows, err := res.RowsAffected()
-			if err != nil {
-				return errors.Wrap(err, "checking for metric rollbacks")
-			}
-			rollbacks[table] = int(affectedRows)
-			if affectedRows > 0 {
-				rollbackHappened = true
+		// CHECK: we skip rollbacks, summary metrics, and total batch updates for generic metrics.
```

Comment:
> * we should do rollbacks for these: trials reporting generic metrics can also have failures, in which case it makes sense to do a rollback.

would we roll back all metric types? let's chat [here](https://hpe-aiatscale.slack.com/archives/C04S3PARXFC/p1683665081907409)

> we should compute summary metrics: it's a useful feature

I agree but leave it off of this PR to avoid adding to the scope?